dask/dataframe/tests/test_groupby.py

```diff
@@ -553,3 +558,143 @@ def test_groupby_multiprocessing():
+
+
+@pytest.mark.parametrize('spec', [
+    'sum', 'min', 'max', 'count', 'size',
+    'std', # NOTE: for std the result is not recast ot the original dtype
+    pytest.mark.xfail(reason="pandas recast to original type")('var'),
+    pytest.mark.xfail(reason="pandas recast to original type")('mean')
+])
+def test_aggregate__single_element_groups(spec):
+    pdf = pd.DataFrame({'a': [1, 1, 3, 3],
+                        'b': [4, 4, 16, 16],
+                        'c': [1, 1, 4, 4],
+                        'd': [1, 1, 3, 3]},
+                       columns=['c', 'b', 'a', 'd'])
+    ddf = dd.from_pandas(pdf, npartitions=3)
+
+    assert_eq(pdf.groupby(['a', 'd']).agg(spec),
+              ddf.groupby(['a', 'd']).agg(spec))
+
+
+def test_aggregate_build_agg_args__reuse_of_intermediates():
+    """Aggregate reuses intermediates. For example, with sum, count, and mean
+    the sums and counts are only calculated once accross the graph and reused to
+    compute the mean.
+    """
+    from dask.dataframe.groupby import _build_agg_args
+
+    no_mean_spec = [
+        ('foo', 'sum', 'input'),
+        ('bar', 'count', 'input'),
+    ]
+
+    with_mean_spec = [
+        ('foo', 'sum', 'input'),
+        ('bar', 'count', 'input'),
+        ('baz', 'mean', 'input'),
+    ]
+
+    no_mean_chunks, no_mean_aggs, no_mean_finalizers = _build_agg_args(no_mean_spec)
+    with_mean_chunks, with_mean_aggs, with_mean_finalizers = _build_agg_args(with_mean_spec)
+
+    assert len(no_mean_chunks) == len(with_mean_chunks)
+    assert len(no_mean_aggs) == len(with_mean_aggs)
+
+    assert len(no_mean_finalizers) == len(no_mean_spec)
+    assert len(with_mean_finalizers) == len(with_mean_spec)
+
+
+def test_aggregate__dask():
+    dask_holder = collections.namedtuple('dask_holder', ['dask'])
+    get_agg_dask = lambda obj: dask_holder({
+        k: v for (k, v) in obj.dask.items() if k[0].startswith('aggregate')
+    })
+
+    specs = [
+        {'b': {'c': 'mean'}, 'c': {'a': 'max', 'a': 'min'}},
+        {'b': 'mean', 'c': ['min', 'max']},
+        ['sum', 'mean', 'min', 'max', 'count', 'size', 'std', 'var'],
+        'sum', 'mean', 'min', 'max', 'count', 'std', 'var',
+
+        # NOTE: the 'size' spec is special since it bypasses aggregate
+        #'size'
+    ]
+
+    pdf = pd.DataFrame({'a': [1, 2, 3, 1, 1, 2, 4, 3, 7] * 100,
+                        'b': [4, 2, 7, 3, 3, 1, 1, 1, 2] * 100,
+                        'c': [0, 1, 2, 3, 4, 5, 6, 7, 8] * 100,
+                        'd': [3, 2, 1, 3, 2, 1, 2, 6, 4] * 100},
+                       columns=['c', 'b', 'a', 'd'])
+    ddf = dd.from_pandas(pdf, npartitions=100)
+
+    for spec in specs:
+        result1 = ddf.groupby(['a', 'b']).agg(spec, split_every=2)
+        result2 = ddf.groupby(['a', 'b']).agg(spec, split_every=2)
+
+        agg_dask1 = get_agg_dask(result1)
+        agg_dask2 = get_agg_dask(result2)
+
+        core_agg_dask1 = {k: v for (k, v) in agg_dask1.dask.items()
```

Comment:
`map_partitions` used in the finalize step passes the meta dataframe. Therefore, `==` can no longer be used for comparison.